Support left/right mouse wheel in terminal emulators

trygveaa opened the pull request:
This adds support for left/right mouse wheel when running in a terminal emulator, for terminal emulators that supports it (e.g. xterm and urxvt).

The support is the same as in gvim, i.e.:
- When nowrap is set, they will scroll the window sideways.
- They can be mapped with `<ScrollWheelLeft>` and `<ScrollWheelRight>`.

Additionally, they are now supported as mouse events in `:terminal`, when an application running in it enables mouse.

trygveaa committed:
Keep mouse release state in a separate variable

Including the MOUSE_RELEASE bits in mouse_code makes scroll right
always interpreted as a release, because it also has those bits set.
Instead, keep it in a separate variable, and only include it in the
button code for non-wheel events.

trygveaa committed:
Support left/right mouse wheel when running in a terminal

This allows you to scroll sideways with the left/right mouse wheel when
nowrap is set, and to map <ScrollWheelLeft> and <ScrollWheelRight>. This
(obviously) only works in terminal emulators that support it (like xterm
and urxvt).

Note that scrolling sideways only work when vim is compiled with gui
enabled (apparently that affects the terminal version too), but mapping
the keys works either way.

trygveaa committed:
Fix support for right mouse wheel in non-SGR protocols

As mentioned two commits ago, right mouse wheel includes the same bits
as MOUSE_RELEASE. Therefore, we can't know if a right mouse wheel event
is a press or release except with the SGR protocol which encodes if it's
a release or not separately.

trygveaa committed:
Support left/right mouse wheel in :terminal

trygveaa commented on src/testdir/test_termcodes.vim:
```diff
@@ -220,10 +221,23 @@ func Test_1xterm_mouse_wheel()
     call assert_equal(1, line('w0'), msg)
     call assert_equal([0, 7, 1, 0], getpos('.'), msg)
+
+    call MouseWheelRight(1, 1)
```
I wanted to test the window column position too here, like the asserts above for mouse wheel up/down, but I didn't find any function to report the current window column position. Is there any?

vim-ml commented:
Hi,

On Tue, Jun 30, 2020 at 12:33 PM Trygve Aaberge <vim-dev-github@256bit.org>
wrote:

> *@trygveaa* commented on this pull request.
> ------------------------------
>
> In src/testdir/test_termcodes.vim
> <https://github.com/vim/vim/pull/6363#discussion_r447930251>:
>
> > @@ -220,10 +221,23 @@ func Test_1xterm_mouse_wheel()
>      call MouseWheelUp(1, 1)
>      call assert_equal(1, line('w0'), msg)
>      call assert_equal([0, 7, 1, 0], getpos('.'), msg)
> +
> +    call MouseWheelRight(1, 1)
>
> I wanted to test the window column position too here, like the asserts
> above for mouse wheel up/down, but I didn't find any function to report the
> current window column position. Is there any?
>
>
>
Depending on what you are checking for, you can use one of the following
functions:
col(), virtcol(), getpos(), getcurpos(), getmousepos(), screencol(),
screenpos() and wincol()

- Yegappan

trygveaa commented:
> Depending on what you are checking for, you can use one of the following functions: col(), virtcol(), getpos(), getcurpos(), getmousepos(), screencol(), screenpos() and wincol() - Yegappan

Ah, thanks, I found [this post](https://stackoverflow.com/a/20157805) which said to use `1 + virtcol(".") - wincol()`. That seems to work. I've added the asserts to the test.

trygveaa commented:
Hm, I see that the tests I added are failing on freebsd and mac, while they are passing on linux. Do you have any tips on debugging this? (I don't have freebsd or mac available)

trygveaa commented:
> Hm, I see that the tests I added are failing on freebsd and mac, while they are passing on linux. Do you have any tips on debugging this? (I don't have freebsd or mac available)

This is fixed now. It turns out that horizontal scrolling is done by the gui feature, even when running in a terminal emulator, and it seems that is disabled in the freebsd and mac tests. So I changed the new tests to only run when gui is enabled.

A further improvement could be to implement horizontal scrolling when the gui feature is disabled.

The two remaining failing builds (huge/gcc-s390x and huge-testgui+coverage/gcc) didn't seem to run properly. I don't think it's related to the changes in this PR.
